src/compiler/checker.ts

```diff
@@ -26185,10 +26197,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const constraint = getConstraintOfTypeParameter(inference.typeParameter);
             if (constraint) {
                 const instantiatedConstraint = instantiateType(constraint, context.nonFixingMapper);
+                // TODO: decide what to do about fallback type
+                if (inferredType && inferredType.flags & TypeFlags.Object && (inferredType as ObjectType).objectFlags & ObjectFlags.ReverseMapped) {
+                    (inferredType as ReverseMappedType).inferenceMapper = context.nonFixingMapper;

```

Comment:
Hm. A mutation like this is odd, especially if the `context.compareTypes` call causes reentrancy on the `inferredType` (different places in the comparison could witness different mappers with different results). Instead, having a non-fixing-mapper clone of the type seems prudent.

Reply:
Ah yes, this mutation is just the best I was able to figure out so far when it comes to exposing this mapper to `inferReverseMappedType`. I meant to comment on that here but you beat me to it and looked here first :P

I could maintain this in some stack or something but OTOH the mutation felt simpler because it just makes it directly accessible at the call site that needs this. It looks odd, especially here, but it's because reverse mapped types just don't participate in the same code path as everything else - all of them are kinda going through their own inference passes, even though they somewhat originate in this "parent" pass (they are not linked to it though). 

I was also thinking that perhaps instead of assigning a mapper I could just assign the `instantiatedConstraint`. In away, the instantiation that I do now in `inferReverseMappedType` is quite similar to potential `getIndexedAccessType(instantiatedConstraint, propertyNameType)`. Would that alleviate the reentrancy concern at all?

> Instead, having a non-fixing-mapper clone of the type seems prudent.

I'm afraid that I don't understand what should be cloned here and where the clone should be used. Could you elaborate on this?

--- 

One extra thing that I was thinking about. `getInferredType` can return cached but non-fixed results. Does it mean that by doing this thing here we might accidentally~ "fix" the properties of a non-fixed result? Or is it just not a concern because the reverse mapped object with those fixed properties doesn't survive `inferredType` clears so it would be recreated and its properties would get "fixed" again (but now with the new `nonFixingMapper`/`instantiatedConstraint`). I didn't yet have time to investigate how this behaves in practice.

Reply:
> I'm afraid that I don't understand what should be cloned here and where the clone should be used. Could you elaborate on this?

Have a copy of the reverse mapped type that stores the `nonFixingMapper` instead of the normal mapper (and probably have that reverse mapped type mutually reference the normal version of the type). (And include that variant-ness in the ID of the reverse mapped type, so they're recognized as separate types in the caches)

> Or is it just not a concern because the reverse mapped object with those fixed properties doesn't survive inferredType clears so it would be recreated and its properties would get "fixed" again (but now with the new nonFixingMapper/instantiatedConstraint). I didn't yet have time to investigate how this behaves in practice.

Yes but no. Reverse mapped types are made in inference, so they *can* be pretty short-lived, but they're identified structurally - all reverse mapped types with the same source, target, and constraint types are the same reverse mapped types, even when they occur at differing places in inference. So even if it's initially short lived, it can be pulled back out later by another attempt to construct one over the same types. That's actually another issue with this as it currently stands - including the mapper in the type (...by including it in the properties via this honestly roundabout way - passing it into `inferTypeForHomomorphicMappedType`/`createReverseMappedType` should clarify things) should affect the type id, but getting a good structural id out of a mapper isn't simple (since function mappers exist aplenty). This means in this PR right now, reverse mapped types mangle one another - only the first inference context within which a reverse mapped type of a certain structure is used gets cached, which is not correct. I think that since reverse mapped types are made within an inference context, them keeping a reference to said context is probably reasonable, but we'll definitely need to check that potentially duplicating reverse mapped types like that isn't terrible for perf on real projects (since inference contexts (and mappers) are rather throwaway).